### PR TITLE
[WIP] Multithread downloading directory

### PR DESCRIPTION
**Pull request by hariton27sy**
### Database name
Databases with physical backups

# Pull request description
Second part of working with directories. Downloader

### CODE REVIEW

File: internal/bundle.go [hunk 1]
```diff
@@ -110,5 +108,23 @@ func (bundle *Bundle) FinishComposing() (TarFileSets, error) {
 }
 
 func (bundle *Bundle) getFileRelPath(fileAbsPath string) string {
-	return utility.PathSeparator + utility.GetSubdirectoryRelativePath(fileAbsPath, bundle.Directory)
+	return utility.GetSubdirectoryRelativePath(fileAbsPath, bundle.Directory)
+}
+
+func (bundle *Bundle) createTarFileInfoHeader(path string, info os.FileInfo) (header *tar.Header, err error) {
+	var symLinkTarget string
+	if info.Mode()&os.ModeSymlink == os.ModeSymlink {
```
Comment (usernamedt) [inline internal/bundle.go:116]: WAL-G default behavior is not to follow symlinks.

File: internal/bundle.go [hunk 2]
```diff
@@ -81,16 +81,14 @@ func (bundle *Bundle) AddToBundle(path string, info os.FileInfo, err error) erro
...
 	if err != nil {
-		return errors.Wrap(err, "addToBundle: could not grab header info")
+		return err
 	}
 
-	fileInfoHeader.Name = bundle.getFileRelPath(path)
 	tracelog.DebugLogger.Println(fileInfoHeader.Name)
 
-	if bundle.FilesFilter.ShouldUploadFile(path) {
+	if bundle.FilesFilter.ShouldUploadFile(path) && info.Mode().IsRegular() {
```
Comment (usernamedt) [inline internal/bundle.go:91]: WAL-G uploads all files, including dirs and symlinks.

**Review by usernamedt — CHANGES_REQUESTED** [review]

File: internal/bundle.go [hunk 3]
```diff
@@ -110,5 +108,15 @@ func (bundle *Bundle) FinishComposing() (TarFileSets, error) {
 }
 
 func (bundle *Bundle) getFileRelPath(fileAbsPath string) string {
-	return utility.PathSeparator + utility.GetSubdirectoryRelativePath(fileAbsPath, bundle.Directory)
+	return utility.GetSubdirectoryRelativePath(fileAbsPath, bundle.Directory)
```
Comment (usernamedt) [inline internal/bundle.go:111]: Why to remove the leading path separator?

File: internal/directory_downloader.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+type DirectoryIsNotEmptyError struct {
+	error
+}
+
+type CommonDirectoryDownloader struct {
+	Folder     storage.Folder
+	BackupName string
+}
+
+func NewCommonDirectoryDownloader(folder storage.Folder, backupName string) DirectoryDownloader {
```
Comment (usernamedt) [inline internal/directory_downloader.go:24]: I guess it is better to pass the explicit list of tars for download to the downloader constructor. Sometimes we don't need to download all the tars from the folder.

Comment (hariton27sy) [reply internal/directory_downloader.go:24]: If you have explicit list of tars then you can use `ExtractAll` function from package `internal`

File: internal/bundle.go (hunk 3 above)
Comment (hariton27sy) [reply internal/bundle.go]: Make behaviour like in postgres relative path

**Review by usernamedt — APPROVED** [review]
LGTM, the only thing left is to fix the failing tests

**Pull request merged**